src/uki.rs

```diff
@@ -0,0 +1,434 @@
+            '"' => loop {
+                result.push(match iter.next()? {
+                    // Strictly speaking, we should only handle \" \$ \` and \\...
+                    '\\' => iter.next()?,
+                    '"' => break,
+                    other => other,
+                });
+            },
+
+            '\'' => loop {
+                result.push(match iter.next()? {
+                    '\'' => break,
+                    other => other,
+                });
+            },
+
+            // Per POSIX we should handle '\\' sequences here, but os-release spec says we'll only
+            // encounter A-Za-z0-9 outside of quotes, so let's not bother with that for now...
+            other => result.push(other),
+        }
+    }
+
+    Some(result)
+}
+
+/// Gets the value that matches the first key that's present and successfully dequotes, or None.
+fn get_value(map: &HashMap<&str, &str>, keys: &[&str]) -> Option<String> {
+    keys.iter()
+        .find_map(|key| map.get(key).and_then(|v| dequote(v)))
+}
+
+/// Gets an appropriate label for display in the boot menu for the given UKI image, according to
+/// the "Type #2 EFI Unified Kernel Images" section in the Boot Loader Specification.  This will be
+/// based on the "PRETTY_NAME" and "VERSION_ID" fields found in the os-release file (falling back
+/// to "ID" and/or "VERSION" if they are not present).
+///
+/// For more information, see:
+///  - https://uapi-group.org/specifications/specs/boot_loader_specification/
+///  - https://www.freedesktop.org/software/systemd/man/latest/os-release.html
+///
+/// # Arguments
+///
+///  * `image`: the complete UKI image as a byte slice
+///
+/// # Return value
+///
+/// If we could successfully parse the provided UKI as a Portable Executable file and find an
+/// ".osrel" section in it, return a string to use as the boootloader entry.  If we were unable to
+/// find any meaningful content in the os-release information this will be "Unknown 0".
+///
+/// If we couldn't parse the PE file or couldn't find an ".osrel" section then an error will be
+/// returned.
+pub fn get_boot_label(image: &[u8]) -> Result<String, UkiError> {
+    let osrel = get_osrel_section(image).ok_or(UkiError::PortableExecutableError)??;
+    let map = HashMap::from_iter(
+        osrel
+            .lines()
+            .filter(|line| !line.trim().starts_with('#'))
+            .filter_map(|line| line.split_once('=')),
+    );
+
+    // At least one of the name fields must be present
+    let mut result = get_value(&map, &["PRETTY_NAME", "NAME", "ID"]).ok_or(UkiError::NoName)?;
+
+    // But version is optional
+    if let Some(version) = get_value(&map, &["VERSION_ID", "VERSION"]) {
+        result.push_str(&format!(" {version}"));
+    }
+
+    Ok(result)
+}
+
+#[cfg(test)]
+mod test {
+    use core::mem::size_of;
+
+    use similar_asserts::assert_eq;
+    use zerocopy::IntoBytes;
+
+    use super::*;
+
+    #[test]
+    fn test_dequote() {
```

Comment:
Optional followup: I think this would make sense as a separate `mod shell_parser` or something?

Reply:
Maybe, but it really does parse a subset of shell as is present in os-release files.  If we need some general-purpose shell parsing stuff in the future I'd consider `shlex`...